sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/FilterUtils.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.sql.SqlOperator;
+import org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.sql.parser.SqlParser;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.expressions.Evaluator;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expression.Operation;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.types.Type.TypeID;
+import org.apache.iceberg.util.DateTimeUtil;
+import org.apache.iceberg.util.NaNUtil;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Utilities that convert between a SQL filter expression and an Iceberg {@link Expression}. Uses
+ * Apache Calcite semantics.
+ *
+ * <p>Note: Only supports top-level fields (i.e. cannot reference nested fields).
+ */
+class FilterUtils {
+  private static final Map<SqlKind, Operation> FILTERS =
+      ImmutableMap.<SqlKind, Operation>builder()
+          .put(SqlKind.IS_NULL, Operation.IS_NULL)
+          .put(SqlKind.IS_NOT_NULL, Operation.NOT_NULL)
+          .put(SqlKind.LESS_THAN, Operation.LT)
+          .put(SqlKind.LESS_THAN_OR_EQUAL, Operation.LT_EQ)
+          .put(SqlKind.GREATER_THAN, Operation.GT)
+          .put(SqlKind.GREATER_THAN_OR_EQUAL, Operation.GT_EQ)
+          .put(SqlKind.EQUALS, Operation.EQ)
+          .put(SqlKind.NOT_EQUALS, Operation.NOT_EQ)
+          .put(SqlKind.IN, Operation.IN)
+          .put(SqlKind.NOT_IN, Operation.NOT_IN)
+          .put(SqlKind.AND, Operation.AND)
+          .put(SqlKind.OR, Operation.OR)
+          .build();
+
+  static FilterFunction filterOn(Expression expression, Schema schema) {
+    return new FilterFunction(expression, schema);
+  }
+
+  static class FilterFunction {
+    private final Evaluator evaluator;
+
+    FilterFunction(Expression expression, Schema schema) {
+      this.evaluator = new Evaluator(schema.asStruct(), expression);
+    }
+
+    boolean filter(Record rec) {
+      return evaluator.eval(rec);
+    }
+  }
+
+  static Expression convert(@Nullable String filter, Schema schema) {
+    if (filter == null) {
+      return Expressions.alwaysTrue();
+    }
+
+    SqlParser parser = SqlParser.create(filter);
+    try {
+      SqlNode expression = parser.parseExpression();
+      return convert(expression, schema);
+    } catch (Exception exception) {
+      throw new RuntimeException(
+          String.format("Encountered an error when parsing filter: '%s'", filter), exception);
+    }
+  }
+
+  private static Expression convert(SqlNode expression, Schema schema) throws SqlParseException {
```

Comment:
Any chance Iceberg itself ships with such a converter?

Reply:
I searched through the code base looking for one but couldn't find one :(
Flink [1] and Spark [2] implement their own conversions too.

[1] https://github.com/apache/iceberg/blob/main/flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/FlinkFilters.java#L75
[2] https://github.com/apache/iceberg/blob/main/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java#L111

Reply:
OK cool, makes sense. Extra credit project: write to dev@iceberg.apache.org pointing to ours, Flink's, and Spark's code that converts Calcite into Iceberg expression, and suggest they adopt it.

Reply:
Sent a message over to their dev slack chat ([thread](https://apache-iceberg.slack.com/archives/C03LG1D563F/p1746547273367289))

Reply:
Yeah, it's unfortunate that all frameworks that use Iceberg has to implement their own converter.